data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/strategy/UnorderedConsumerRecordOffsetStrategy.java

```diff
@@ -57,64 +71,106 @@ public UnorderedConsumerRecordOffsetStrategy(final KafkaConsumer<?, ?> consumer,
   }
 
   /**
    * {@inheritDoc}
    */
   @Override
   public void recordDiscarded(final KafkaConsumerRecord<?, ?> record) {
-    successfullySentToSubscriber(record);
+    mutateStateAndCheckAck(new TopicPartition(record.topic(), record.partition()), record.offset());
+  }
+
+  /**
+   * @return null if it shouldn't ack, otherwise the offset to ack - 1.
+   */
+  private Long mutateStateAndCheckAck(TopicPartition topicPartition, long offset) {
+    long lastAckedOffset = this.lastAckedPerPartition.get(topicPartition); // This is always non null
+    SortedSet<Long> partitionPendingAcks =
+      this.pendingAcksPerPartition.computeIfAbsent(topicPartition, v -> new TreeSet<>());
+    partitionPendingAcks.add(offset);
+
+    // Return the ack to commit if:
+    // * last acked offset is the same of the first pending ack in the set
+    // * the set contains every element in its range of values
+    return (partitionPendingAcks.first() == lastAckedOffset + 1 &&
+      partitionPendingAcks.last() - partitionPendingAcks.first() == partitionPendingAcks.size() - 1) ?
+      partitionPendingAcks.last() : null;
   }
 
-  private Future<Map<TopicPartition, OffsetAndMetadata>> commit(
-    final KafkaConsumerRecord<?, ?> record) {
-    logger.debug("committing record {}", record);
-    return consumer.commit(Map.of(
-      new TopicPartition(record.topic(), record.partition()),
-      new OffsetAndMetadata(record.offset() + 1, ""))
-    );
+  private void commit(final KafkaConsumerRecord<?, ?> record) {
+    TopicPartition topicPartition = new TopicPartition(record.topic(), record.partition());
+    Long toAck = mutateStateAndCheckAck(topicPartition, record.offset());
+    if (toAck != null) {
+      // Reset the state
+      long lastAckedBeforeThisOne = this.lastAckedPerPartition.put(topicPartition, toAck);
+      SortedSet<Long> messagesImGoingToAck = this.pendingAcksPerPartition.remove(topicPartition);
+
+      // Execute the actual commit
+      consumer.commit(Map.of(
+        topicPartition,
+        new OffsetAndMetadata(toAck + 1, ""))
+      ).onSuccess(ignored -> {
+        eventsSentCounter.increment(messagesImGoingToAck.size());
+        logger.debug(
+          "committed {} {} {}",
+          keyValue("topic", record.topic()),
+          keyValue("partition", record.partition()),
+          keyValue("offset", toAck + 1)
+        );
+      })
+        .onFailure(cause -> {
+          // If the commit failed, there are 2 situations:
+          // * Somebody committed with an offset greater than this one, so we just discard the error
+          // * Nobody tried to commit again, so let's restore the state
+          if (!(this.lastAckedPerPartition.get(topicPartition) > toAck)) {
+            this.lastAckedPerPartition.put(topicPartition, lastAckedBeforeThisOne);
+            this.pendingAcksPerPartition.compute(topicPartition, (k, actual) -> {
+              if (actual != null) {
```

Comment:
Is this possible since we call `mutateStateAndCheckAck` above?

Reply:
It is, because somebody could have asked meanwhile to add a new commit and that state could have been populated again